### PR TITLE
Embedded dirk

### PR DESCRIPTION
**Pull request by sconde**
Adds embedded  DIRK capability to Tempus.

@trilinos/tempus 

## Motivation and Context

Add adaptive time stepping capability.

## Related Issues

* Closes 
* Blocks 
* Is blocked by #2219 
* Follows 
* Precedes 
* Related to 
* Part of 
* Composed of 

## How Has This Been Tested?

All tests passing. DIRK test has been updated to include embedded DIRK.
[Wednesday-2018-02-14-16_25-build_tempus_dev-test.log](https://github.com/trilinos/Trilinos/files/1725995/Wednesday-2018-02-14-16_25-build_tempus_dev-test.log)

## Checklist

- [ ] My commit messages mention the appropriate GitHub issue numbers.
- [ ] My code follows the code style of the affected package(s).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

### CODE REVIEW

File: packages/tempus/src/Tempus_StepperExplicitRK_impl.hpp [hunk 1]
```diff
@@ -298,7 +342,13 @@ StepperExplicitRK<Scalar>::getValidParameters() const
   //            << "  RK Explicit 2 Stage 2nd order by Runge\n"
   //            << "  RK Explicit Trapezoidal\n";
 
-  return ERK_ButcherTableau_->getValidParameters();
+  Teuchos::RCP<Teuchos::ParameterList> pl = Teuchos::parameterList();
+  pl->set<bool>("Use Embedded", false,
+    "'Whether to use Embedded Stepper (if available) or not"
```
Comment (rppawlo) [inline packages/tempus/src/Tempus_StepperExplicitRK_impl.hpp]: need a \n on line 347

File: packages/tempus/src/Tempus_StepperExplicitRK_impl.hpp [hunk 2]
```diff
@@ -307,7 +357,11 @@ Teuchos::RCP<Teuchos::ParameterList>
 StepperExplicitRK<Scalar>::getDefaultParameters() const
 {
   Teuchos::RCP<Teuchos::ParameterList> pl = Teuchos::parameterList();
-  *pl = *(ERK_ButcherTableau_->getValidParameters());
+  pl->set<bool>("Use Embedded", false,
+    "'Whether to use Embedded Stepper (if available) or not"
```
Comment (rppawlo) [inline packages/tempus/src/Tempus_StepperExplicitRK_impl.hpp]: same thing here - need a newline

File: packages/tempus/src/Tempus_TimeStepControl_impl.hpp [hunk 3]
```diff
@@ -19,10 +19,13 @@
 #include "Tempus_TimeStepControlStrategyConstant.hpp"
 #include "Tempus_TimeStepControlStrategyComposite.hpp"
 #include "Tempus_TimeStepControlStrategyBasicVS.hpp"
+#include "Tempus_TimeStepControlStrategyIntegralController.hpp"
 
 //Thyra
 #include "Thyra_VectorStdOps.hpp"
 
+//TODO: remove this
+#include <fstream>
```
Comment (rppawlo) [inline packages/tempus/src/Tempus_TimeStepControl_impl.hpp]: did you mean to leave this in?

File: packages/tempus/test/DIRK/Tempus_DIRK_VanDerPol.xml [hunk 4]
```diff
@@ -101,7 +156,117 @@
...
+     <ParameterList name="Default Solver">
+        <ParameterList name="NOX">
+           <ParameterList name="Direction">
+              <Parameter name="Method" type="string" value="Newton"/>
+              <ParameterList name="Newton">
+                 <Parameter name="Forcing Term Method" type="string" value="Constant"/>
+                 <Parameter name="Rescue Bad Newton Solve" type="bool" value="1"/>
+              </ParameterList>
+           </ParameterList>
+           <ParameterList name="Line Search">
```
Comment (rppawlo) [inline packages/tempus/test/DIRK/Tempus_DIRK_VanDerPol.xml]: This is quite a long list of options. Is it possible to get away with more defaults to trim down the input files? I think you could remove the options, line search and direction sublists from the nox solver.

**Comment by allevin** [comment]
@sconde Error from build 261

In file included from /projects/sems/install/rhel6-x86_64/sems/compiler/gcc/4.9.3/openmpi/1.6.5/include/mpi.h:253:0,
                 from /scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_gcc_4.9.3/Trilinos/packages/teuchos/core/src/Teuchos_Time.hpp:56,
                 from /scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_gcc_4.9.3/Trilinos/packages/teuchos/comm/src/Teuchos_TimeMonitor.hpp:70,
                 from /scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_gcc_4.9.3/Trilinos/packages/tempus/test/ExplicitRK/Tempus_ExplicitRK_FSA.hpp:11,
                 from /scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_gcc_4.9.3/Trilinos/packages/tempus/test/ExplicitRK/Tempus_ExplicitRK_Combined_FSA.cpp:9:
/projects/sems/install/rhel6-x86_64/sems/compiler/gcc/4.9.3/openmpi/1.6.5/include/mpi_portable_platform.h:374:34: warning: invalid suffix on literal; C++11 requires a space between literal and string macro [-Wliteral-suffix]
              _STRINGIFY(__GNUC__)"."_STRINGIFY(__GNUC_MINOR__)"."_STRINGIFY(__GNUC_PATCHLEVEL__)
                                  ^
/projects/sems/install/rhel6-x86_64/sems/compiler/gcc/4.9.3/openmpi/1.6.5/include/mpi_portable_platform.h:374:63: warning: invalid suffix on literal; C++11 requires a space between literal and string macro [-Wliteral-suffix]
              _STRINGIFY(__GNUC__)"."_STRINGIFY(__GNUC_MINOR__)"."_STRINGIFY(__GNUC_PATCHLEVEL__)
                                                               ^
/scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_gcc_4.9.3/Trilinos/packages/tempus/test/ExplicitRK/Tempus_ExplicitRKTest.cpp: In member function ‘virtual void Tempus_Test::ExplicitRK_EmbeddedVanDerPol_UnitTest::runUnitTestImpl(Teuchos::FancyOStream&, bool&) const’:
/scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_gcc_4.9.3/Trilinos/packages/tempus/test/ExplicitRK/Tempus_ExplicitRKTest.cpp:380:13: error: redeclaration of ‘const int refIstep’
   const int refIstep = 61;
             ^
/scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_gcc_4.9.3/Trilinos/packages/tempus/test/ExplicitRK/Tempus_ExplicitRKTest.cpp:377:13: note: ‘const int refIstep’ previously declared here
   const int refIstep = 47;
             ^
[ 81%] Linking CXX executable Tempus_ForwardEuler.exe
[ 81%] Built target Tempus_ForwardEuler
Scanning dependencies of target Tempus_Integrator
[ 81%] Building CXX object packages/tempus/test/Integrator/CMakeFiles/Tempus_Integrator.dir/Tempus_IntegratorTest.cpp.o
make[2]: *** [packages/tempus/test/ExplicitRK/CMakeFiles/Tempus_ExplicitRK.dir/Tempus_ExplicitRKTest.cpp.o] Error 1
make[1]: *** [packages/tempus/test/ExplicitRK/CMakeFiles/Tempus_ExplicitRK.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

**Comment by sconde** [comment]
Thank you so much for your review and comments.

> In TimeStepControl::setTimeStepControlStrategy() line 443,
> why do we need to check if this paramter is available?
> "tscPL_->isParameter("Amplification Factor") and"
> A comment mention for backward compatibility, but shouldn't this just be
> a new TimeStepControlStrategy? I would suggest we not support two versions
> since this is a fairly new strategy to begin with.

Applications that had not updated their Tempus `Time Step Control` sublist to use the `Time Step Control Strategy` would have had issues with running variable time stepping. I saw this with Drekar when I was making the changes. This piece of code can be deleted altogether if everyone changes their Tempus PL to use the `Time Step Control Strategy` and include a sublist for the `Basic VS` strategy when `Integrator Step Type =  Variable`.

**Comment by ccober6** [comment]
@ikalash @sconde So the problems that this should affect are Albany, Drekar and Tempus. Can we make the changes necessary to drop this backward compatibility case?  I guess there are two ways we could go.  Remove this code and break Albany, Drekar and Tempus, forcing us to fix it now.  Or, we could proceed with the current backward compatibiilty code and start another issue, and allow time for us to fix it and then remove the code.  I am in favor of the first, as the second will likely be drawn out.  Thoughts?

**Comment by ikalash** [comment]
@ccober6 : I am OK with your proposed approach, which would break Albany.  It sounds like the fix would be pretty straightforward.  I can see which tests break in Albany from our CDash once this change to Tempus is made and fix them at that time.  Or, maybe a better approach, is for me to build the branch that has the fix, build Albany against it, fix all the issues; then when the branch is merged into develop, commit my Albany changes.

**Comment by sconde** [comment]
@ccober6 @ikalash @rppawlo @eric-c-cyr This could also break Empire-Fluid

**Comment by sconde** [comment]
@ccober6 Would you mind looking over this once again, please

**Pull request merged**